### ISSUE TITLE
bool keys fail to target bool axes, even when explicitly specified

### ISSUE THREAD

**Issue by gdementen** [larray-project/larray#735]
```python
>>> arr = LArray([0.1, 0.2], Axis([False, True], 'gender'))
>>> key = np.array([True, False, True, True])
>>> arr[key]
ValueError: boolean key with a different shape ((4,)) than array ((2,))
>>> arr.points[key]
ValueError: boolean key with a different shape ((4,)) than array ((2,))
```

Oddly, using list keys seems to work (as long as we only target a single dimension)
```python
>>> arr[[True, False, True, True]]
gender  True  False  True  True
         0.2    0.1   0.2   0.2
>>> arr.points[[True, False, True, True]]
gender  True  False  True  True
         0.2    0.1   0.2   0.2
```

Using an LArray key does not seem to help either:

```python
>>> arr[LArray(key, Axis(6, 'gender'))]
ValueError: gender[0, 2, 3] is not a valid label for any axis
```

This is a "regression" in 0.29. It is only a partial regression because it was known to be broken before. It's just more broken now. 😄 See liam2/liam2#296.

**Comment by gdementen** [comment]
tentatively adding to 0.30. Will push to 0.31 if necessary, but this should be done sooner rather than later.

**Comment by benjello** [comment]
@gdementen : when do you think you could look into this. I am stucked in a dependence nightmare. I want to compile liam2/master in a Ptyhon3.7. liam2 requires larray=0.28 which requires older version of pandas which cannot be installed Cython/numpy problems. I would like to use the latest version of pandas.

**Comment by gdementen** [comment]
Ouch. Sorry about this mess, but this will take a while. Certainly not before mid march and probably not before the end of may and even that I cannot promise because LIAM2 is not the only project I had to put aside for a while. FWIW, I am spending most of my time these days on the website which will hold the results of the evaluation of political parties programs that the FPB is doing for the belgian elections in May. I am truly sorry about this situation but I cannot help it. 

If you are desperate, you could use larray 0.28 with a backport of https://github.com/larray-project/larray/commit/ca21828f097631fa3ba990e612858e20a0fdf529 to add support for Pandas >= 0.21.

**Comment by benjello** [comment]
I am desperate. Thanks for the hint !

**Comment by benjello** [comment]
It worked :-)
Could I push a branch called ipp on this repo for this backport ? Since I need to ask people from a secured data center to download it for me I would prefer to do the cheery-pick here than there. The fact that it is the official repo may also make it feel safer.

**Comment by gdementen** [comment]
Sure.

**Comment by gdementen** [comment]
Note that manually specifying the axis does not work either:

```python
>>> arr[X.gender[key]]
IndexError: boolean index did not match indexed array along dimension 0; dimension is 2 but corresponding boolean dimension is 4
>>> arr[arr.gender[key]]
IndexError: boolean index did not match indexed array along dimension 0; dimension is 2 but corresponding boolean dimension is 4
```

At first I **thought** it worked because I was inadvertently using the list workaround:

```python
>>> arr[arr.gender[True, False, True, True]]
gender  True  False  True  True
         0.2    0.1   0.2   0.2
>>> arr[X.gender[True, False, True, True]]
gender  True  False  True  True
         0.2    0.1   0.2   0.2
```
(but this was all a lie)

**Comment by gdementen** [comment]
This last part was easily fixable (I just fixed it in my working copy). The first part is less so. Not because it is technically difficult but because it is ambiguous and we have to decide what to do. It is ambiguous because the boolean array can refer to either filtering the array "values" and thus **filter** (potentially combined) axes via a positional filter (True, False, True, True would thus take the [first, third and fourth] labels of an hypothetical axis) _OR_ **select** one or several axis labels from a boolean axis ([True, False, True, True] would translate to [1, 0, 1, 1] meaning [second, first, second, second]). 

I am leaning towards not doing anything more in LArray (than what I just fixed) and try to "fix" this in LIAM2 (if required and possible) and require the user to explicitly name/select the axis when one wants to target a boolean axis in normal LArray code. The reasoning is as follow:
* a numpy array is usually considered equivalent to an LArray with anonymous wildcard axes (this is what is converted to anyway)
* arr[bool_array] will transform the bool_array into a tuple of igroups and it would be confusing if I the presence of a boolean axis changed the behavior. I mean it would be **more** confusing than having boolean axes behave differently than other axes.
* boolean axes should be rare in practice. I know that we currently have a ton in LIAM2 code. This is all because we lack support for "string" (or better yet: categorical) fields. String fields would be hard to implement given the limited support in numexpr (and would be inefficient anyway) but categorical should not be too hard and would be great to have anyway.

So, this is for the long term picture, but I need to come up with a workable solution **now** (I can't wait to have implemented categoricals in LIAM2).
